src/SFML/Graphics/Shader.cpp

```diff
@@ -156,11 +210,50 @@ Shader::CurrentTextureType Shader::CurrentTexture;
 
 
 ////////////////////////////////////////////////////////////
+struct Shader::UniformBinder : private sf::NonCopyable
+{
+    ////////////////////////////////////////////////////////////
+    /// \brief Constructor: set up state before uniform is set
+    ///
+    ////////////////////////////////////////////////////////////
+    UniformBinder(Shader* shader, const std::string& name) :
+    program(0),
+    location(-1)
+    {
+        if (shader->m_shaderProgram)
+        {
+            ensureGlContext();
+
+            // Enable program
+            glCheck(program = GLEXT_glGetHandle(GLEXT_GL_PROGRAM_OBJECT));
+            glCheck(GLEXT_glUseProgramObject(castToGlHandle(shader->m_shaderProgram)));
```

Comment:
That sounds smart, I can try that. 

There seems to be other optimization potential, e.g. by caching `setUniform()` calls, maybe in a future PR... Ironically, this would again necessitate a design based on functors/virtual functions :wink: